File: org.ai4fm.proofprocess.project.core/src/org/ai4fm/proofprocess/project/core/ProofManager.java
package org.ai4fm.proofprocess.project.core;

import java.io.IOException;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

import org.ai4fm.proofprocess.log.ProofLog;
import org.ai4fm.proofprocess.log.ProofProcessLogFactory;
import org.ai4fm.proofprocess.project.Project;
import org.ai4fm.proofprocess.project.ProjectProofProcessFactory;
import org.eclipse.core.resources.IFile;
import org.eclipse.core.resources.IProject;
import org.eclipse.core.runtime.CoreException;
import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.core.runtime.NullProgressMonitor;
import org.eclipse.core.runtime.QualifiedName;
import org.eclipse.core.runtime.SubProgressMonitor;
import org.eclipse.core.runtime.jobs.Job;
import org.eclipse.emf.common.command.BasicCommandStack;
import org.eclipse.emf.common.notify.AdapterFactory;
import org.eclipse.emf.common.util.URI;
import org.eclipse.emf.ecore.EObject;
import org.eclipse.emf.ecore.resource.Resource;
import org.eclipse.emf.ecore.xmi.XMLResource;
import org.eclipse.emf.ecore.xmi.impl.XMIResourceFactoryImpl;
import org.eclipse.emf.edit.domain.AdapterFactoryEditingDomain;
import org.eclipse.emf.edit.provider.ComposedAdapterFactory;


public class ProofManager {
	
	public static final String PROOF_FOLDER = ".proofprocess/";
	private static final String PROOF_PROJECT_PATH = PROOF_FOLDER + "project.proof";
	private static final String PROOF_LOG_PATH = PROOF_FOLDER + "project.prooflog";
	
	public static final Map<String, Object> SAVE_OPTIONS;
	static {
		Map<String, Object> opts = new HashMap<String, Object>();
		// do not use XML 1.1, because its implementation in Java 1.6 has bugs
		// regarding large attributes due to old Xerces libraries.
//		opts.put(XMLResource.OPTION_XML_VERSION, "1.1");
		opts.put(XMLResource.OPTION_ENCODING, "UTF-8");
		SAVE_OPTIONS = Collections.unmodifiableMap(opts);
	}
	
	/**
	 * Key for the loaded project reference on resource.
	 */
	public final static QualifiedName PROP_PROOF_PROJECT = 
			new QualifiedName(ProjectProofProcessPlugin.PLUGIN_ID, "proofProject"); //$NON-NLS-1$
	
	/**
	 * Key for the loaded proof log reference on resource.
	 */
	public final static QualifiedName PROP_PROOF_LOG = 
			new QualifiedName(ProjectProofProcessPlugin.PLUGIN_ID, "proofLog"); //$NON-NLS-1$
	
	public static Project getProofProject(IProject projectResource, IProgressMonitor monitor)
			throws CoreException {
		
		return getProofModel(projectResource, PROP_PROOF_PROJECT, monitor);
	}
	
	public static ProofLog getProofLog(IProject projectResource, IProgressMonitor monitor)
			throws CoreException {
		
		return getProofModel(projectResource, PROP_PROOF_LOG, monitor);
	}
	
	private static <T extends EObject> T getProofModel(IProject projectResource,
			QualifiedName prop, IProgressMonitor monitor) throws CoreException {
		
		Object modelObj = projectResource.getSessionProperty(prop);
		if (modelObj == null) {
			loadProofModelsJob(projectResource, monitor);
			modelObj = projectResource.getSessionProperty(prop);
		}
		
		@SuppressWarnings("unchecked")
		T model = (T) modelObj;
		return model;
	}
	
	public static void loadProofModelsJob(IProject projectResource, IProgressMonitor monitor)
			throws CoreException {
		
		if (monitor == null) {
			monitor = new NullProgressMonitor();
		} else {
			monitor = new SubProgressMonitor(monitor, 10);
		}
		
		try {
			
			Job.getJobManager().beginRule(projectResource, monitor);
			
			monitor.beginTask("Loading proof progress", IProgressMonitor.UNKNOWN);
			
			loadProofModels(projectResource);
			
		} finally {
			Job.getJobManager().endRule(projectResource);
			monitor.done();
		}
	}
	
	private static void loadProofModels(IProject projectRes) 
			throws CoreException {
		
		// check maybe they has already been loaded (double-checked locking)
		Project proj = (Project) projectRes.getSessionProperty(PROP_PROOF_PROJECT);
		ProofLog log = (ProofLog) projectRes.getSessionProperty(PROP_PROOF_LOG);
		
		if (proj != null && log != null) {
			// loaded
			return;
		}
		
		initResourceFactories();
		
		// Create a ComposedAdapterFactory for all registered models
		AdapterFactory adapterFactory = new ComposedAdapterFactory(
				ComposedAdapterFactory.Descriptor.Registry.INSTANCE);
		AdapterFactoryEditingDomain editingDomain = new AdapterFactoryEditingDomain(
				adapterFactory, new BasicCommandStack());
		
		Project proofProject = loadProofFile(editingDomain, getProofProjectFile(projectRes), 
				ProjectProofProcessFactory.eINSTANCE.createProject());
		
		ProofLog proofLog = loadProofFile(editingDomain, getProofLogFile(projectRes), 
				ProofProcessLogFactory.eINSTANCE.createProofLog());
		
		projectRes.setSessionProperty(PROP_PROOF_PROJECT, proofProject);
		projectRes.setSessionProperty(PROP_PROOF_LOG, proofLog);
	}
	
	private static <T extends EObject> T loadProofFile(AdapterFactoryEditingDomain editingDomain, IFile file, T empty) {
		
		URI uri = URI.createPlatformResourceURI(file.getFullPath().toString(), true);
		Resource emfResource = editingDomain.getResourceSet().createResource(uri);
		
		T root = null;
		
		if (file.exists()) {
			try {
				emfResource.load(null);
			} catch (IOException e) {
				ProjectProofProcessPlugin.log(e);
			}
			
			@SuppressWarnings("unchecked")
			T elem = (T) emfResource.getContents().get(0); 
			root = elem;
		}
		
		if (root == null) {
			// unable to load - init a new model file
			root = empty;
			emfResource.getContents().add(root);
		}
		
		return root;
	}
	
	private static IFile getProofProjectFile(IProject projectResource) {
		return projectResource.getFile(PROOF_PROJECT_PATH);
	}
	
	private static IFile getProofLogFile(IProject projectResource) {
		return projectResource.getFile(PROOF_LOG_PATH);
	}
	
	private static void initResourceFactories() {
		
		// Register the XMI resource factory for the .proof and .prooflog extensions
		Resource.Factory.Registry reg = Resource.Factory.Registry.INSTANCE;
		reg.getExtensionToFactoryMap().put("proof", new XMIResourceFactoryImpl());
		reg.getExtensionToFactoryMap().put("prooflog", new XMIResourceFactoryImpl());
	}

}

File: org.ai4fm.proofprocess.project.core/src/org/ai4fm/proofprocess/project/core/ProofManager.scala
package org.ai4fm.proofprocess.project.core

import java.io.IOException

import scala.collection.JavaConverters._

import org.ai4fm.proofprocess.cdo.PProcessCDO
import org.ai4fm.proofprocess.log.{ProofLog, ProofProcessLogFactory}
import org.ai4fm.proofprocess.project.{Project, ProjectProofProcessFactory}
import org.ai4fm.proofprocess.project.core.internal.ProjectPProcessCorePlugin.{error, log, plugin}

import org.eclipse.core.filesystem.URIUtil
import org.eclipse.core.resources.{IFile, IProject}
import org.eclipse.core.runtime.{
  CoreException,
  IProgressMonitor,
  NullProgressMonitor,
  QualifiedName,
  SubProgressMonitor
}
import org.eclipse.core.runtime.jobs.Job
import org.eclipse.emf.cdo.CDOObject
import org.eclipse.emf.cdo.transaction.CDOTransaction
import org.eclipse.emf.cdo.util.CommitException
import org.eclipse.emf.common.command.BasicCommandStack
import org.eclipse.emf.common.util.{EList, URI}
import org.eclipse.emf.ecore.EObject
import org.eclipse.emf.ecore.resource.Resource
import org.eclipse.emf.ecore.xmi.impl.XMIResourceFactoryImpl
import org.eclipse.emf.edit.domain.AdapterFactoryEditingDomain
import org.eclipse.emf.edit.provider.ComposedAdapterFactory


/**
 * @author Andrius Velykis
 */
object ProofManager {

  /** Key for the loaded project reference on resource */
  lazy private val PROP_PROOF_PROCESS =
    new QualifiedName(plugin.pluginId, "proofProcess") //$NON-NLS-1$

  @throws(classOf[CoreException])
  def proofProject(projectResource: IProject,
                   monitor: IProgressMonitor = new NullProgressMonitor): Project =
    proofManager(projectResource, monitor).proofs

  @throws(classOf[CoreException])
  def proofLog(projectResource: IProject,
               monitor: IProgressMonitor = new NullProgressMonitor): ProofLog =
    proofManager(projectResource, monitor).proofLog
  
  @throws(classOf[CoreException])
  private def proofManager(projectResource: IProject, monitor: IProgressMonitor): ProjectProofManager = {

    // retrieve the stored manager, or load one otherwise
    storedManager(projectResource) getOrElse {

      // create/load manager
      val submonitor = subMonitor(monitor)

      try {
        // start a rule on the resource to block when loading the manager
        Job.getJobManager.beginRule(projectResource, monitor)

        // check maybe the manager has already been loaded (double-checked locking)
        storedManager(projectResource) getOrElse {

          monitor.beginTask("Initialising proof manager", IProgressMonitor.UNKNOWN)

          // TODO what if the project gets renamed?
          val projectProofMan = loadProofProcess(projectResource, monitor)

          projectResource.setSessionProperty(PROP_PROOF_PROCESS, projectProofMan)

          projectProofMan
        }

      } finally {
        Job.getJobManager.endRule(projectResource)
        submonitor.done()
      }
    }
  }

  private def loadProofProcess(projectResource: IProject, monitor: IProgressMonitor): ProjectProofManager = {

    // Create a ComposedAdapterFactory for all registered models
    val adapterFactory = new ComposedAdapterFactory(ComposedAdapterFactory.Descriptor.Registry.INSTANCE)
    val editingDomain = new AdapterFactoryEditingDomain(adapterFactory, new BasicCommandStack)

    // FIXME better repository names?
    val repositoryName = projectResource.getName

    // store database in workspace plug-in location
    // TODO investigate configuration/user locations
    // (via Platform.getConfigurationLocation or .getUserLocation)
    val databaseLocPath = plugin.getStateLocation.append("database")
    val databaseLocUri = URIUtil.toURI(databaseLocPath)

    val session = PProcessCDO.session(databaseLocUri, repositoryName, monitor)
    val transaction = session.openTransaction(editingDomain.getResourceSet)

    // resolve the "proof" and "prooflog" resources in the CDO repository
    val proofRoot = loadCreateMigrateRoot(transaction, projectResource,
      "proof",
      proofProjectFile(projectResource),
      ProjectProofProcessFactory.eINSTANCE.createProject,
      {
        case p: Project => Some(p)
        case _ => None
      },
      monitor)

    val proofLogRoot = loadCreateMigrateRoot(transaction, projectResource,
      "prooflog",
      proofLogFile(projectResource),
      ProofProcessLogFactory.eINSTANCE.createProofLog,
      {
        case l: ProofLog => Some(l)
        case _ => None
      },
      monitor)

    ProjectProofManager(proofRoot, proofLogRoot)
  }

  private def storedManager(projectResource: IProject): Option[ProjectProofManager] =
    projectResource.getSessionProperty(PROP_PROOF_PROCESS) match {
      case manager: ProjectProofManager => Some(manager)
      case _ => None
    }

  private def subMonitor(monitor: IProgressMonitor): IProgressMonitor = {
    Option(monitor) match {
      case Some(monitor) => new SubProgressMonitor(monitor, 10)
      case None => new NullProgressMonitor
    }
  }

  private def loadCreateMigrateRoot[T <: EObject](transaction: CDOTransaction, projectResource: IProject,
    repositoryName: String, migrateFile: IFile, create: => T, convert: EObject => Option[T],
    monitor: IProgressMonitor): T = {

    // resolve the resource in the CDO repository
    val proofResource = loadResource(transaction, repositoryName)
    val proofRoot = proofResource.asScala.headOption.flatMap(convert) match {
      case Some(p) => p
      case None => {
        // the root project is not available - allocate a new one
        // create a new project, add to the resource and commit

        // before creating a new one, try migrating existing XML-based model
        val migrateProject = migrateXmlFileHistory(projectResource, migrateFile, convert, monitor)

        val newProject = migrateProject getOrElse create

        proofResource.clear
        proofResource.add(newProject)
        commitTransaction(transaction, monitor)

        newProject
      }
    }

    proofRoot
  }

  private def loadResource(transaction: CDOTransaction, resourceName: String): EList[EObject] = {
    val emfResource = transaction.getOrCreateResource(resourceName)

    try {
      // load EMF resource
      emfResource.load(null)
    } catch {
      case e: IOException => log(error(Some(e)))
    }

    emfResource.getContents
  }

  def commitTransaction(cdoObj: CDOObject, monitor: IProgressMonitor) {
    cdoObj.cdoView match {
      case tr: CDOTransaction => commitTransaction(tr, monitor)
      case _ => log(error(msg = Some("Cannot commit transaction - object does not belong to one: " + cdoObj)))
    }
  }

  def commitTransaction(transaction: CDOTransaction, monitor: IProgressMonitor) {
    try {
      transaction.commit(monitor)
    } catch {
      case e: CommitException => log(error(Some(e)))
    }
  }

  private case class ProjectProofManager(val proofs: Project, val proofLog: ProofLog)

  /** Load file history from XML to migrate */
  private def migrateXmlFileHistory[T](projectResource: IProject, file: IFile, convert: EObject => Option[T],
    monitor: IProgressMonitor): Option[T] = {

    // Create a ComposedAdapterFactory for all registered models
    val adapterFactory = new ComposedAdapterFactory(ComposedAdapterFactory.Descriptor.Registry.INSTANCE)
    val editingDomain = new AdapterFactoryEditingDomain(adapterFactory, new BasicCommandStack)

    val proofRoot = loadProofFile(editingDomain, file)

    proofRoot.flatMap(convert)
  }

  private def loadProofFile[T](editingDomain: AdapterFactoryEditingDomain, file: IFile): Option[EObject] = {

    val uri = URI.createPlatformResourceURI(file.getFullPath().toString(), true)
    val emfResource = editingDomain.getResourceSet().createResource(uri)

    if (file.exists) {
      try {
        // load EMF resource
        emfResource.load(null)
      } catch {
        case e: IOException => log(error(Some(e)))
      }

      emfResource.getContents.asScala.headOption
    } else {
      None
    }
  }

  def proofFolder = ".proofprocess/"

  private def proofProjectFile(projectResource: IProject) =
    projectResource.getFile(proofFolder + "project.proof")

  private def proofLogFile(projectResource: IProject) =
    projectResource.getFile(proofFolder + "project.prooflog")

  private def initResourceFactories() {

    // Register the XMI resource factory for the .proof and .prooflog extensions
    val reg = Resource.Factory.Registry.INSTANCE;
    reg.getExtensionToFactoryMap().put("proof", new XMIResourceFactoryImpl());
    reg.getExtensionToFactoryMap().put("prooflog", new XMIResourceFactoryImpl());
  }

}
